### ISSUE TITLE
Array pass-by-reference does not work in GPU locale

### ISSUE THREAD

**Issue by xianghao-wang** [chapel-lang/chapel#22114]
### Summary of Problem
This issue is raised when try to fixing `NotGPU.chpl` by replacing `const` with `param` in  #22113. 
If a function called in GPU locale and this function changes the value of an array passed by the parameter, the array values do not change.

### Steps to Reproduce

**Source Code:**
`ErrorGPU.chpl`
```chapel
use Random;
use AutoMath;
use GPU;

param N = 16;
param B = 4;

param P = 15.0;

proc main() {
  on here.gpus[0] {
    var A: [0..<N] real;
    fillRandom(A);

    writeln(A);
    foreach i in 0..<N / B {
      add(A, i);
    }
    writeln(A);
  }
  
}

proc add(A: [] real, group: int) {
  const ind = group * B;
  for i in 0..<B {
    A[ind+i] = P;
  }
}
```

**Compile command:**
`chpl --fast --no-ieee-float ErrorGPU.chpl`

**Execution command:**
`./ErrorGPU`

**Output:**
```
0.377858 0.319206 0.089764 0.455717 0.996789 0.335745 0.0327119 0.0454723 0.953064 0.144297 0.463564 0.590515 0.529496 0.148187 0.312785 0.931109
0.377858 0.319206 0.089764 0.455717 0.996789 0.335745 0.0327119 0.0454723 0.953064 0.144297 0.463564 0.590515 0.529496 0.148187 0.312785 0.931109
```

**Note:**
By removing `on here.gpus[0]`, the program works perfectly on CPU.

**Comment by e-kayrakli** [comment]
Thanks @xianghao-wang! This is definitely strange. I suspect it may be the same underlying issue as https://github.com/chapel-lang/chapel/issues/21961#issuecomment-1483198423.

**Comment by DanilaFe** [comment]
A smaller instance of this issue is the following:

```Chapel
use GPU;

param N = 16;

proc main() {
  on here.gpus[0] {
    var A: [0..<N] real;

    writeln(A);
    foreach i in 0..<N {
      add(A, i);
    }
    writeln(A);
  }
}

proc add(ref A: [] real, idx: int) {
  A[idx] = 1;
}
```
Even though `add(A, i)` is called to set `A[i]` to 1, this prints all zeroes.

**Comment by DanilaFe** [comment]
Alright, so making `add` inline seems to make the code work.

<details>
<summary>Technical details here!</summary>

Making two copies of `add`, one inline and one non-inline, I'm observing that the kernel generated for calling an inline function (which works) takes as one of its arguments the `_ddata` of an array, and simply performs assignment. 

```
function chpl_gpu_kernel[2274545](in arg _literal_1[2274577]:int(64)[13], in arg _ic__F1_high[2274586]:int(64)[13], in arg coerce_tmp[2274637]:_ddata(real(64))[1440493]) : _unknown[54] "codegen for GPU" "always resolve function" "resolved"
  {
    (index checking)

    ref call_tmp[2274635]:real(64)[16] "dead after last mention" "maybe param" "temp"
    (2274639 'move' call_tmp[2274635](2274641 'array_get' coerce_tmp[2274637] chpl_simt_index[2274579]))
    const-val coerce_tmp[2274644]:real(64)[16] "dead after last mention" "coerce temp" "insert auto destroy" "temp"
    (2274646 'move' coerce_tmp[2274644](2274648 'cast' real(64)[16] 1))
    (2274651 '=' call_tmp[2274635] coerce_tmp[2274644])
    label _end_chpl_gpu_kernel[2274629]
    (2274631 return _void[58])
    (2274547 'local block')
  }
```

On the other hand, the non-inline version of the kernel takes in the whole `array` struct, and tries to perform unpacking / `instance` computations there, via the non-inline `add`.

```
  function chpl_gpu_kernel[2274682](in arg _literal_1[2274714]:int(64)[13], in arg _ic__F1_high[2274723]:int(64)[13], in arg A[2274772]:_array(unmanaged DefaultRectangularArr(1,int(64),false,real(64),int(64)))[1445804]) : _unknown[54] "codegen for GPU" "always resolve function" "resolved"
  {
    (index computations)

    (2274775 call add2[1488891] A[2274772] chpl_simt_index[2274716])
    label _end_chpl_gpu_kernel[2274766]
    (2274768 return _void[58])
    (2274684 'local block')
  }

  function add2[1488891](const  ref arg A[1488893]:_array(unmanaged DefaultRectangularArr(1,int(64),false,real(64),int(64)))[1445804], const in arg idx[1488895]:int(64)[13]) : void[4] "compiler added where" "codegen for GPU" "codegen for CPU and GPU" "instantiated generic" "invisible fn" "resolved" "fn synchronization free" "no return value for void"
  {
    val ret[2234286]:DefaultRectangularArr(1,int(64),false,real(64),int(64))[1441391] "dead after last mention" "RVV" "temp"
    (2234288 'move' ret[2234286](2234290 '.v' A[1488893] _instance[1445798]))
    val coerce_tmp[2234293]:DefaultRectangularArr(1,int(64),false,real(64),int(64))[1441391] "dead after last mention" "coerce temp" "insert auto destroy" "temp"
    (2234295 'move' coerce_tmp[2234293] ret[2234286])
    val coerce_tmp[2234310]:_ddata(real(64))[1440493] "dead after last mention" "coerce temp" "insert auto destroy" "temp"
    (2234312 'move' coerce_tmp[2234310](2234314 '.v' coerce_tmp[2234293] shiftedData[1441367]))
    ref call_tmp[2234317]:real(64)[16] "dead after last mention" "maybe param" "temp"
    (2234319 'move' call_tmp[2234317](2234321 'array_get' coerce_tmp[2234310] idx[1488895]))
    const-val coerce_tmp[1488957]:real(64)[16] "dead after last mention" "coerce temp" "insert auto destroy" "temp"
    (1488963 'move' coerce_tmp[1488957](1488959 'cast' real(64)[16] 2))
    (2172540 '=' call_tmp[2234317] coerce_tmp[1488957])
    (1488938 return _void[58])
  }
```
</details>

**Comment by e-kayrakli** [comment]
Probably it'd be interesting to think about whether we are doing the right thing while passing an array argument to a kernel. To save you 30 seconds:

- https://github.com/chapel-lang/chapel/blob/main/compiler/codegen/cg-expr.cpp#L5129-L5146 is where we decide how to pass arguments to a kernel. Where one option is just to directly pass it, the other is to pass it by offload. The former will pass a 0 alongside the symbol, the latter will pass the size of the thing to be offloaded. I think we should offload the record that represent the array. But that might have been the wrong call on my part (I implemented this part waay waaay back).
- https://github.com/chapel-lang/chapel/blob/main/runtime/src/gpu/cuda/gpu-cuda.c#L241-L266 is where we use those 0s and non-0s to actually pass args to the kernel.

---

It'd be also curious to see:

```chapel
use GPU;

param N = 16;

proc main() {
  on here.gpus[0] {
    var A: [0..<N] real = 1;
    var B: [0..<N] real;

    writeln(A);
    foreach i in 0..<N {
      add(A, i);
      B[i] = A[i];
    }
    writeln(A);
    writeln(B); // do we really get all 1s here, or is it junk?
  }
}

proc add(ref A: [] real, idx: int) {
  A[idx] = 1;
}
```

**Comment by DanilaFe** [comment]
What's curious is that it seems like the address of the array's buffer is the same in all cases, even the non-inline case where the array doesn't end up getting updated...

```
use GPU;
use CTypes;

param N = 16;

proc main() {
  on here.gpus[0] {
    var A: [0..<N] real;
    var P: [0..0] int;

    writeln(A);
    writeln(c_ptrTo(A) : int);

    foreach i in 0..<N {
        P[0] = add1(A, i);
        assertOnGpu();
    }
    writeln(A);
    writeln(P);
    P[0] = 0;

    foreach i in 0..<N {
        P[0] = add2(A, i);
        assertOnGpu();
    }
    writeln(A);
    writeln(P);
    P[0] = 0;
  }
}

inline proc add1(ref A: [] real, idx: int) {
  A[idx] = 1;
  return c_pointer_return(A[0]) : int;
}

proc add2(ref A: [] real, idx: int) {
  A[idx] = 2;
  return c_pointer_return(A[0]) : int;
}
```

```
0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0
139736895196160
Arg: 0, size: 0
Arg: 1, size: 0
Arg: 2, size: 0
Arg: 3, size: 32
1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0
139736895196160
Arg: 0, size: 0
Arg: 1, size: 0
Arg: 2, size: 0
Arg: 3, size: 32
1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0 1.0
139736895196160
```

**Comment by DanilaFe** [comment]
> do we really get all 1s here, or is it junk?

I'm not sure what you meant by junk, but writing out into B from the kernel seems to show that A isn't even change inside the kernel.

```Chapel
use GPU;

param N = 16;

proc main() {
  on here.gpus[0] {
    var A: [0..<N] real = 1;
    var B: [0..<N] real;

    writeln(A);
    foreach i in 0..<N {
      add(A, i);
      B[i] = A[i];
    }
    writeln(A);
    writeln(B); // do we really get all 1s here, or is it junk?
  }
}

proc add(ref A: [] real, idx: int) {
  A[idx] = 42;
}
```
Prints only ones, not 42!

**Comment by e-kayrakli** [comment]
Investigating how things are passed to add2 in PTX can turn up something. And maybe try to compare that with an equivalent CUDA code? If PTX isn't clear, SASS must be. But it is the actual assembly and much harder to understand than PTX.

**Comment by DanilaFe** [comment]
We've determined this is caused by the insertion of wide references into local versions of procedures called from the GPU loop. It seems like a bug / oversight in the transformation of these local procedures doesn't correctly use local versions of the variables, which leads to calls to communication from the GPU. Currently, communication is unsupported on the GPU (and we'd much prefer the local versions to be called!). The "unsupported" communication shows up as a no-op in the runtime, so all assignments have no effect.

**Comment by e-kayrakli** [comment]
@xianghao-wang -- This issue was automatically closed after @DanilaFe's #22373. Please let us know if you can still observe the bug on `main` (and reopen the issue if that's the case). The fix is going to be in 1.31 release.
